virtcontainers/podlist.go

```diff
@@ -0,0 +1,91 @@
+//
+// Copyright (c) 2018 HyperHQ Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"fmt"
+	"sync"
+)
+
+type podlist struct {
+	lock sync.RWMutex
+	pods map[string]*Pod
+}
+
+// globalPodlist tracks pods globally
+var globalPodlist = podlist{pods: make(map[string]*Pod)}
+
+func (pl *podlist) addPod(pod *Pod) (err error) {
+	if pod == nil {
+		return nil
+	}
+
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+	if pl.pods[pod.id] == nil {
+		pl.pods[pod.id] = pod
+	} else {
+		err = fmt.Errorf("pod %s exists", pod.id)
+	}
+	return err
+}
+
+func (pl *podlist) removePod(id string) error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+	delete(pl.pods, id)
+	return nil
+}
+
+func (pl *podlist) lookupPod(id string) (pod *Pod, err error) {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+	if pl.pods[id] != nil {
+		pod = pl.pods[id]
+	} else {
+		err = fmt.Errorf("pod %s does not exist", id)
+	}
+	return pod, err
+}
+
+func (pl *podlist) foreachPod(f func(*Pod) error) error {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+	for _, pod := range pl.pods {
+		err := f(pod)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func addPod(pod *Pod) error {
+	return globalPodlist.addPod(pod)
+}
+
+func removePod(id string) error {
+	return globalPodlist.removePod(id)
+}
+
+func lookupPod(id string) (*Pod, error) {
+	return globalPodlist.lookupPod(id)
+}
+
+func foreachPod(f func(*Pod) error) error {
+	return globalPodlist.foreachPod(f)
+}
```

Comment:
I would remove those:
```go
func addPod(pod *Pod) error {
	return globalPodlist.addPod(pod)
}

func removePod(id string) error {
	return globalPodlist.removePod(id)
}

func lookupPod(id string) (*Pod, error) {
	return globalPodlist.lookupPod(id)
}

func foreachPod(f func(*Pod) error) error {
	return globalPodlist.foreachPod(f)
}
```

I think they add more confusion where we can call directly from `pod.go`:
```go
globalPodList.addPod(p)
...
```